clients/pkg/promtail/client/client.go

```diff
@@ -287,14 +467,15 @@ func (c *client) run() {
 			// If adding the entry to the batch will increase the size over the max
 			// size allowed, we do send the current batch and then create a new one
 			if batch.sizeBytesAfter(e) > c.cfg.BatchSize {
-				c.sendBatch(tenantID, batch)
-
-				batches[tenantID] = newBatch(c.maxStreams, e)
+				_ = dispatchBatch(tenantID, batch)
```

Comment:
So two answers:
> we want to ensure the batch has been sent successfully before creating a new one and clearing the previous WAL segment

Sending the batch, and if successful deleting the corresponding segment occurs on the **WAL reading** side, that is the watcher and the configured callbacks. For example, [here's](https://github.com/grafana/loki/blob/05553737f80d0a5dda4de421381ea01aa4fdce0f/clients/pkg/promtail/client/client.go#L779-L785) the piece of code under the `SegmentEnd` callback that sends the batch and if no error occurs, deletes the batch.

In case of failure when sending the batch (considering the `sendBatch` does some retries depending on the client config), the segment won't be deleted and would be picked up by the next Promtail restart.

> we should be checking the error here and later in run as well, should we not?
I think we could be logging the error, but in the writer side if the `dispatch` fails it means that cutting a segment in the WAL fails. Therefore, all follow up entries will be written to the same segment, therefore not lost. Only problem with this is that the batch that corresponds to that segment could keep growing, and once it's read, the send operation failing.

Reply:
Logged error on failed batch `dispatches`